assets/js/modules/analytics-4/datastore/report.js

```diff
@@ -152,6 +156,104 @@ const baseSelectors = {
 
 		return reports[ stringifyObject( options ) ];
 	},
+
+	/**
+	 * Determines whether the Analytics 4 module is still gathering data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean|undefined} Returns `true` if gathering data, otherwise `false`. Returns `undefined` while resolving.
+	 */
+	isGatheringData: createRegistrySelector( ( select ) => () => {
+		const { startDate, endDate } = select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+		} );
+
+		const args = {
+			dimensions: [ 'date' ],
+			metrics: [ { name: 'totalUsers' } ],
+			startDate,
+			endDate,
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
+
+		// Disable reason: select needs to be called here or it will never run.
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const report = select( MODULES_ANALYTICS_4 ).getReport( args );
+		const hasResolvedReport = select(
+			MODULES_ANALYTICS_4
+		).hasFinishedResolution( 'getReport', [ args ] );
+
+		if ( ! hasResolvedReport ) {
+			return undefined;
+		}
+
+		if ( ! isPlainObject( report ) ) {
+			return false;
+		}
+
+		if ( ! Array.isArray( report?.rows ) || report?.rows?.length === 0 ) {
+			return true;
+		}
+
+		return false;
+	} ),
+
+	/**
+	 * Determines whether Analytics 4 has zero data or not.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean|undefined} Returns FALSE if not gathering data and the report is not zero, otherwise TRUE. If the request is still being resolved, returns undefined.
+	 */
+	hasZeroData: createRegistrySelector( ( select ) => () => {
+		const { startDate, endDate } = select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+		} );
+
+		const args = {
+			dimensions: [ 'date' ],
+			metrics: [ { name: 'totalUsers' } ],
+			startDate,
+			endDate,
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
+
+		const isGatheringData = select( MODULES_ANALYTICS_4 ).isGatheringData();
+		if ( isGatheringData === undefined ) {
+			return undefined;
+		}
+
+		// Disable reason: select needs to be called here or it will never run.
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const report = select( MODULES_ANALYTICS_4 ).getReport( args );
+		const hasResolvedReport = select(
+			MODULES_ANALYTICS_4
+		).hasFinishedResolution( 'getReport', [ args ] );
```

Comment:
I believe we can return early if we know that we are still gathering data, can't we?

```js
const isGatheringData = select( MODULES_ANALYTICS_4 ).isGatheringData();
if ( isGatheringData === undefined ) {
    return undefined;
}

if ( isGatheringData === true ) {
    return true;
}

const report = select( MODULES_ANALYTICS_4 ).getReport( args );
...
```

Reply:
Although it's correct in isolation, I don't think we should make this change. The reason being, this code is copied from the UA `isGatheringData()` with very minor changes. If and when we need to apply a change to the logic in both of these `isGatheringData()` selectors, it will be more difficult if they have diverged unnecessarily.

In fact, we have a concrete example of this in the pipeline as can be seen in https://github.com/google/site-kit-wp/pull/6565.

So, I think this is best left as-is for now.

Reply:
My apologies, I thought this was in the `isGatheringData()` selector.

It seems logical to me that if we're copying the logic from one piece of code then it's easier to maintain if both of those pieces of code remain in sync, for future changes that should be applied to both of them.

However, as this is in `hasZeroData()` my comment about changes in the pipeline doesn't apply, so I'd be happy to apply this change here and to the UA version.